### PR TITLE
feat(dbt): allow dbt tests to be modeled as asset checks

### PR DESCRIPTION
**Pull request by rexledesma**
## Summary & Motivation
As the title.

Currently, users will have to opt-in to modelling their dbt tests as Dagster asset checks. This is because the API's for asset checks are currently experimental, and we do not want to break users of the public stable API's for ingesting dbt models as Dagster software-defined assets.

So, to enable this feature, users will have to configure the following metadata on their dbt project tests:

```yaml
# dbt_project.yml
tests:
  +meta:
    dagster:
      asset_check: True
```

Or, tests can be individually turned on using the dbt config block or property file: https://docs.getdbt.com/reference/test-configs#test-specific-configurations.

## How I Tested These Changes
pytest

### CODE REVIEW

**Comment by rexledesma** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16324** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16324" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16324?utm_source=stack-comment).

File: python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py [hunk 1]
```diff
@@ -262,13 +275,14 @@ def is_successful(self) -> bool:
         return self.process.wait() == 0
 
     @public
-    def stream(self) -> Iterator[Union[Output, AssetObservation]]:
+    def stream(self) -> Iterator[Union[Output, AssetObservation, AssetCheckResult]]:
         """Stream the events from the dbt CLI process and convert them to Dagster events.
 
         Returns:
-            Iterator[Union[Output, AssetObservation]]: A set of corresponding Dagster events.
+            Iterator[Union[Output, AssetCheckResult]]: A set of corresponding Dagster events.
```
Comment (johannkm) [inline python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py:282]: should still have observation for now right

**Review by johannkm — APPROVED** [review]
Nice, I'm surprised how straightforwards this is

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py [hunk 2]
```diff
@@ -509,6 +511,27 @@ def default_description_fn(dbt_resource_props: Mapping[str, Any], display_raw_sq
...
+
+
+def default_asset_check_fn(
+    asset_key: AssetKey, dbt_resource_props: Mapping[str, Any]
+) -> Optional[AssetCheckSpec]:
+    is_asset_check = is_asset_check_from_dbt_resource_props(dbt_resource_props)
+    if not is_asset_check:
+        return None
+
+    severity = AssetCheckSeverity(dbt_resource_props["config"]["severity"])
```
Comment (johannkm) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py:525]: we'll need to reconcile with https://github.com/dagster-io/dagster/pull/16320

**Pull request merged**